saw-remote-api/src/SAWServer/Data/Contract.hs

```diff
@@ -0,0 +1,87 @@
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE OverloadedStrings #-}
+module SAWServer.Data.Contract
+  ( Contract(..)
+  , ContractVar(..)
+  , Allocated(..)
+  , PointsTo(..)
+  ) where
+
+import Data.Aeson (FromJSON(..), withObject, (.:))
+import Data.Text
+
+import SAWServer
+import SAWServer.Data.SetupValue ()
+
+data Contract ty cryptolExpr =
+  Contract
+    { preVars       :: [ContractVar ty]
+    , preConds      :: [cryptolExpr]
+    , preAllocated  :: [Allocated ty]
+    , prePointsTos  :: [PointsTo cryptolExpr]
+    , argumentVals  :: [CrucibleSetupVal cryptolExpr]
+    , postVars      :: [ContractVar ty]
+    , postConds     :: [cryptolExpr]
+    , postAllocated :: [Allocated ty]
+    , postPointsTos :: [PointsTo cryptolExpr]
+    , returnVal     :: Maybe (CrucibleSetupVal cryptolExpr)
+    }
+    deriving (Functor, Foldable, Traversable)
+
+data ContractVar ty =
+  ContractVar
+    { contractVarServerName :: ServerName
+    , contractVarName       :: Text
+    , contractVarType       :: ty
+    }
+
+data Allocated ty =
+  Allocated
+    { allocatedServerName :: ServerName
+    , allocatedType       :: ty
+    , allocatedMutable    :: Bool
+    , allocatedAlignment  :: Maybe Int
+    }
+
+data PointsTo cryptolExpr =
+  PointsTo
+    { pointer  :: CrucibleSetupVal cryptolExpr
+    , pointsTo :: CrucibleSetupVal cryptolExpr
+    } deriving (Functor, Foldable, Traversable)
+
+instance FromJSON cryptolExpr => FromJSON (PointsTo cryptolExpr) where
```

Comment:
There's an implicit description of a data format in the protocol here. This should be added to `docs/Protocol.rst` eventually, if this design seems mostly ready to go. We haven't written the SAW part up yet due to flux, but it seems to be settling down at least a bit.

Reply:
Yeah, let's write that up soon. I've created #84 to track that.